### PR TITLE
feat: add `upcast` expression pt. 2

### PR DESCRIPTION
**Pull request by jaschdoc**
Related to #4264

### CODE REVIEW

**Comment by jaschdoc** [comment]
It seems like the purity and effect set is still not inferred from the context. Rejecting any program with purity and effect variables would also reject a valid program like this. Any ideas?
```flix

    @test
    def testUpcast01(): Unit =
        let _ =
            if (true)
                upcast x -> x + 1
            else
                x -> { println("x"); x + 1 };
        ()
```
with error
```
-- Safety Error -------------------------------------------------- main/test/flix/Test.Exp.Upcast.flix

>> The following upcast is unsafe and not allowed.

7 |                 upcast x -> x + 1
                    ^^^^^^^^^^^^^^^^^
                    the upcast occurs here.

Actual type  : Int32 -> Int32
Actual purity: true
Actual effect: Empty

Expected type  : Int32 -> Int32 & Impure
Expected purity: b86490
Expected effect: e86491
```

**Comment by magnus-madsen** [comment]
Ah yes. I think we should just ignore the pvar and evar for now and focus just on the type.

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala [hunk 1]
```diff
@@ -181,7 +181,13 @@ object Safety {
       visitExp(exp)
 
     case Expression.Upcast(exp, tpe, pur, eff, loc) =>
-      throw InternalCompilerException("Not implemented")
+      val es = (tpe, pur, eff) match {
+        case (_: Type.Var, _, _) => List(UnsafeUpcast(exp.tpe, exp.pur, exp.eff, tpe, pur, eff, loc))
+        case (_, _: Type.Var, _) => List(UnsafeUpcast(exp.tpe, exp.pur, exp.eff, tpe, pur, eff, loc))
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Safety.scala:186]: So allow this case.

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala [hunk 2]
```diff
@@ -181,7 +181,13 @@ object Safety {
...
+        case (_, _, _: Type.Var) => List(UnsafeUpcast(exp.tpe, exp.pur, exp.eff, tpe, pur, eff, loc))
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Safety.scala:187]: And this case.

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala [hunk 3]
```diff
@@ -180,8 +180,67 @@ object Safety {
     case Expression.Cast(exp, _, _, _, _, _, _, _) =>
       visitExp(exp)
 
-    case Expression.Upcast(exp, tpe, pur, eff, loc) =>
-      throw InternalCompilerException("Not implemented")
+    case e@Expression.Upcast(exp, tpe, pur, eff, loc) =>
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Safety.scala:183]: This looks like a lot of fun :) :)

But try write a recursive function on that matches on `expected` and `actual` simultaneously. 

You will find this a lot easier, I think.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Safety.scala]: Like this:

```
def isSafeUpcast(expected: Type, actual: Type): Bool = (expected, actual) match {
  case (Type.Cst(...), Type.Cst(...)) => // do checks
  case (Type.Apply(...), Type.Apply(...) => // recurse on the parts
/// and so forth
```

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Safety.scala]: So you line them up one at a time

**Comment by jaschdoc** [comment]
Am I on the right track with this?

**Comment by magnus-madsen** [comment]
> Am I on the right track with this?

Not really no 😆 Maybe we can talk tomorrow :)

**Comment by jaschdoc** [comment]
Ok I had some help from @mlutze so this should be good for the purity checks.

**Comment by mlutze** [comment]
$e_1 < e_2 \iff \exists e_3 . (e_1 \wedge e_3) \equiv_B e_2$

Is that true and should we use it?

**Comment by jaschdoc** [comment]
RFR

**Pull request merged**